public/app/plugins/datasource/grafana-azure-monitor-datasource/resource_picker/resource_picker_data.ts

```diff
@@ -0,0 +1,168 @@
+    takes data formated like this:
+    [{Subscription},{Resource Group}] or like this:[{Resource}]
+    and transforms to data that looks like this:
+    { 
+      subscriptionId: {
+        ...subscription data
+
+        children: {
+          resourceGroupName: {
+            ...resource group data
+
+            children: {
+              resourceId: {
+                ...resource data
+              }
+            }
+          }
+        }
+      }
+    }
+  */
+  _formatAndCacheResourceData(rawData: any[]) {
+    // "produce" from immer takes in an object, allows you to edit a copy of that object and then returns a new object
+    this._formattedResourcePickerData = produce(this._formattedResourcePickerData, (draftState: any) => {
+      rawData.forEach((item: any) => {
+        switch (item.type) {
+          case 'microsoft.resources/subscriptions':
+            draftState[item.subscriptionId] = {
+              // handles potential duplicates (see note on proccessing resource groups before subscriptions)
+              ...(draftState[item.subscriptionId] ? draftState[item.subscriptionId] : {}),
+
+              name: item.name,
+              id: item.subscriptionId,
+              subscriptionId: item.subscriptionId,
+              typeLabel: 'Subscription',
+              type: EntryType.Collection,
+              children: {},
+            };
+
+            break;
+
+          case 'microsoft.resources/subscriptions/resourcegroups':
+            // handles processing resourcegroup before we encounter it's parent subscription
+            if (!draftState[item.subscriptionId]) {
+              draftState[item.subscriptionId] = {
+                children: {},
+              };
+            }
+
+            // a note here that we store resource groups by name and not by id because the resources we fetch only have resource group names not ids
+            draftState[item.subscriptionId].children[item.name.toLowerCase()] = {
+              name: item.name,
+              id: item.id,
+              subscriptionId: item.subscriptionId,
+              type: EntryType.SubCollection,
+              typeLabel: 'Resource Group',
+              children: {},
+            };
+            break;
+
+          default:
+            // We assume everything else to be a selectable resource
+            draftState[item.subscriptionId].children[item.resourceGroup].children[item.id] = {
+              name: item.name,
+              id: item.id,
+              type: EntryType.Resource,
+              typeLabel: item.type, // TODO: these types can be quite long, we may wish to format them more
+              location: item.location, // TODO: we may wish to format these locations, by default they are written as 'northeurope' rather than a more human readable "North Europe"
+            };
+        }
+      });
+    });
+
+    return this._formattedResourcePickerData;
+  }
+
+  async _makeResourceGraphRequest(query: string, maxRetries = 1): Promise<any> {
```

Comment:
this PR has some overlap with [this](https://github.com/grafana/grafana/pull/33293/files) which will be merged for beta2. This method (actually maybe everything in this file) should probably be moved to the [arg datasource file](https://github.com/grafana/grafana/blob/2112c3152e1b305ab12853fe7cc3c6e5a072abec/public/app/plugins/datasource/grafana-azure-monitor-datasource/azure_resource_graph/azure_resource_graph_datasource.ts).

Reply:
Yeah I'm not clear on if these requests should be hitting the same endpoints as those featured in that pr or not. 

The data I'm trying to fetch here seems very similar to me from what I would see in that PR so it felt to me like we could use the query method to get back this metadata. But then it sounded to me from Josh that Logs and Metrics can have different client ids that require a [different auth credential](https://github.com/grafana/grafana/issues/33765) and that a "datasource" is a unique concept different from these metadata type of requests? Perhaps we can discuss more on monday

Reply:
Theres is/maybe some overlap between these two things, but it's unclear to me right now what would be the same in practice _in the frontend_.

I think i would like to merge them both in and have a second pass cleaning up (perhaps related to the plugin proxy/auth refactor)